### PR TITLE
Membership term changes

### PR DESCRIPTION
**Pull request by freelock**
Going ahead and putting here for discussion... this moves the ball forward in terms of having a default install, and a bit more UI in place...

### CODE REVIEW

**Comment by bradjones1** [comment]
Thanks for this. A few items, in no particular order:

1. Mind rebasing against `8.x-dev`, since I already committed your toolbar PR? In general, PR's shouldn't include one another, though I've totally done this before ;-)
1. What do you think of using some sort of time-interval field like the one in [Interval module](https://www.drupal.org/project/interval) for the time interval? At the moment it looks like this is just a string, which doesn't lend itself well to programmatic modification. I haven't really vetted Interval per se, but I'd be in favor of using some existing time-management scheme to make this easily parsed.
1. To that end, we need to consider some sort of "indefinite" term as well.
1. I get your motivation for bundling some defaults, but especially if we use some sort of library for times, adding those would be trivial, and avoids installing entities that just get deleted if the user isn't interested in them?
 1. I'll make some comments inline on the code, but in a few cases we should be using dependency injection where you're accessing services off `\Drupal`.

I'll make some other notes in the code, but in general I get where you're going here. While not all implementations will truly have a fixed "term" for the membership (e.g., in my example where the term is indefinite and managed by an external service) this serves more or less the same purpose as my original "Offer" concept, with the addition of a time frame.

File: src/Form/MembershipTermEntityTypeForm.php [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+ */
+class MembershipTermEntityTypeForm extends EntityForm {
+
+  /**
+   * {@inheritdoc}
+   */
+  public function form(array $form, FormStateInterface $form_state) {
+    $form = parent::form($form, $form_state);
+
+    $entity_type_manager = \Drupal::service('entity_type.manager');
```
Comment (bradjones1) [inline src/Form/MembershipTermEntityTypeForm.php:21]: Per my notes in the PR, we should be implementing `::create()` and `_construct()` in the dependency injection model for accessing container services. `EntityForm` extends `FormBase`, which implements `ContainerInjectionInterface`.

File: config/install/field.storage.membership_term_entity.field_term_length.yml [hunk 2]
```diff
@@ -0,0 +1,21 @@
+uuid: a937ea7b-f8e2-4504-96e4-641e94b38154
+langcode: en
+status: true
+dependencies:
+  module:
+    - membership
+id: membership_term_entity.field_term_length
+field_name: field_term_length
+entity_type: membership_term_entity
+type: string
```
Comment (bradjones1) [inline config/install/field.storage.membership_term_entity.field_term_length.yml:10]: Per my notes, thinking we should perhaps use a more structured field type here?

File: membership.workflows.yml [hunk 3]
```diff
@@ -16,3 +16,35 @@ membership_default:
       label: Restart
       from: [expired]
       to: active
+
+membership_term_default:
+  id: membership_term_default
+  label: Default Membership Term Workflow
```
Comment (bradjones1) [inline membership.workflows.yml:22]: This isn't so much a workflow for a membership term as much as a term-limited membership itself. Maybe "Default Term-Limited Membership Workflow" would be clearer?

File: membership.workflows.yml [hunk 4]
```diff
@@ -16,3 +16,35 @@ membership_default:
...
+  group: membership_term
+  states:
+    pending:
+      label: Pending
+    active:
+      label: Active
+    expiring:
+      label: Expiring
```
Comment (bradjones1) [inline membership.workflows.yml:30]: Interesting addition of "Expiring." I don't see any code (yet) that would automate the transition from Active to Expiring (or any other transitions, actually) but is that the idea? Similar with the Pending state... that seems to be an implementation detail, not all sites will have a pending state or need to implement it.

I think the "expiring" state could mean different things to different people, or not apply at all... do you think breaking this out into a second module-provided workflow would make sense? And keep the "default-default" to Active and Expired?

File: src/Entity/MembershipTermEntity.php [hunk 5]
```diff
@@ -8,53 +8,57 @@
...
 /**
- * Defines the Membership Offer entity.
+ * Defines the Membership term entity entity.
  *
  * @ingroup membership
  *
  * @ContentEntityType(
- *   id = "membership_offer",
- *   label = @Translation("Membership Offer"),
+ *   id = "membership_term_entity",
```
Comment (bradjones1) [inline src/Entity/MembershipTermEntity.php:19]: Is there a reason for the `_entity` portion? Most D8 entities just define themselves with a meaningful machine name, I think the `_entity` piece is a little duplicative and implies an additional distinction where there is none?

File: src/Form/MembershipTermEntityDeleteForm.php [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+
+/**
+ * Provides a form for deleting Membership term entity entities.
+ *
+ * @ingroup membership
+ */
+class MembershipTermEntityDeleteForm extends ContentEntityDeleteForm {
+
+
+}
```
Comment (bradjones1) [inline src/Form/MembershipTermEntityDeleteForm.php:15]: Remove excess whitespace

**Review by bradjones1 — CHANGES_REQUESTED** [review]
Some specific code notes.

**Comment by bradjones1** [comment]
I think we need to flesh out the exact purpose of the "Membership Term," as you've worked it up so far.

Simply put, I think there's some conflict in approach between the state of the membership, and the state of the membership term. When I originally wrote the module, the "Membership Offer," which you're replacing, was a one-to-many relationship to memberships which implemented that offer. Basically the membership itself had a state, and the offer represented a particular set of terms or benefits.

You've taken the "Membership Term" concept from the D7 module and brought it here, where (I believe) you expect a 1-to-1 relationship to Memberships. From my read of the D7 `membership_entity` module, the piece that is so far missing is that a Membership is 1-to-many to membership terms, which might represent re-joins, etc. Your PR has a workflow state on the Membership Term but does not remove the state of the Membership - is that purposeful or an oversight? I still think it's the Membership itself that has a state. That state would then change in response to the Membership Term, e.g., one month, and a Membership with a month-long term could be changed to expiration via some sort of audit method (which doesn't appear to exist currently.)

If the Membership Term entity is merely a way of establishing and tracking the time-based term of membership, your note in the issue queue about using inline entity form probably becomes the better approach, and we can remove any of the term's own HTML routes. The Membership Offer entity was built with display in mind because it represented something independent of a specific membership, but a Membership Term doesn't appear to do that.

**Comment by freelock** [comment]
Hi, Brad,

Thanks for the review! My attention has been elsewhere the past couple days, but back on this tonight to Friday. I did already get quite a bit further down the path for our immediate project, so I think I'll be focusing on getting that set up right and then coming back and refactoring for a new PR once I have the basic model in place...

Responses to your specific lists:

>     Mind rebasing against 8.x-dev, since I already committed your toolbar PR? In general, PR's shouldn't include one another, though I've totally done this before ;-)

No problem, though I think this will end up being a couple weeks off (and there will be a lot more in it...)

>     What do you think of using some sort of time-interval field like the one in Interval module for the time interval? At the moment it looks like this is just a string, which doesn't lend itself well to programmatic modification. I haven't really vetted Interval per se, but I'd be in favor of using some existing time-management scheme to make this easily parsed.

My plan was just to use a text field that gets sent directly to strtotime(), which we've used a lot for this purpose. Haven't seen this Interval field, seems like a decent way to make it more clear and provide validation... Will go ahead and incorporate this...

>     To that end, we need to consider some sort of "indefinite" term as well.

Agreed, I was labeling this "Lifetime" in the current code I'm working on...

>     I get your motivation for bundling some defaults, but especially if we use some sort of library for times, adding those would be trivial, and avoids installing entities that just get deleted if the user isn't interested in them?

Perhaps move them to a submodule, "membership starter" or something? I'm finding as I'm developing across 3 different copies of the site, having a config created as soon as I turn the module on extremely helpful -- I'm regularly clobbering local dev copies from the database we're building up, and not enabling membership on that upstream db until the structure is solid.

>     I'll make some comments inline on the code, but in a few cases we should be using dependency injection where you're accessing services off \Drupal.

Agreed, I was just copying some snippets I had opened, at this point running to get a result. I was definitely planning to come back and clean that up, if somebody didn't beat me to it...


> Interesting addition of "Expiring." I don't see any code (yet) that would automate the transition from Active to Expiring (or any other transitions, actually) but is that the idea? Similar with the Pending state... that seems to be an implementation detail, not all sites will have a pending state or need to implement it.

I haven't yet figured out how state_machine supports this, been struggling to find any documentation or patterns here.

Our current project will likely have more states than this, after meeting with the client. We're building out the following:

- Pending -- a membership that is not paid for yet, not active
- Active -- an active membership
- 60- -- active term is expiring within 60 days, and has had the 60-day warning sent
- 30- -- active term is expiring within 30 days, and has had the 30-day warning sent
- 0- -- active term has expired, and has had the expiration warning sent
- Expired -- active term is 30 days past due

... the "60", "30", and "0" states would all be considered "Expiring" -- something specific to this site. With this approach, if payment is made between 60 and 30 days, we can add a new "Pending" membership term with a start date that matches the expiration of the currently active term, and set the membership back to Active, skipping the later notifications. (Haven't gotten here yet, might find a better approach when we do -- any suggestions/better approaches here?)

We see a lot of sites where the "expiration" date does not match up with the actual date the membership gets yanked -- most of our clients want a grace period. So "expiring" represents a membership that has reached some threshold where they can purchase a renewal, but have not yet lost access to the site.

> I think the "expiring" state could mean different things to different people, or not apply at all... do you think breaking this out into a second module-provided workflow would make sense? And keep the "default-default" to Active and Expired?

I'm fine with any defaults we want to ship, and moving more complex workflows to submodules or other contrib modules -- I was thinking having 3 or 4 different workflows available out of the gate, largely differing by the renewal process -- things like what is the start date of a renewal term if they renew 2 weeks after it expired: today, or the expiration date of the previous term? Different organizations want different behavior here...

### Other notes

Your other feedback, I don't have any other comment on, will incorporate...

### Other changes

... I've been shuffling around a lot of methods between the Membership and MembershipType entities, same between MembershipTerm and MembershipTermType, trying to figure out the best place to put each method. And I've ended up creating a MembershipManager service instead for most of the actions.

So I'm seeing that the MembershipManager service (membership.manager) can retrieve the memberships for a particular user, optionally of a particular MembershipType, and handle most of the interactions around this.

Then I'm thinking of different MembershipType and MembershipTermType classes that can be extended for particular business rules, all adhering to the same MembershipTypeInterface. The Service will just load up the appropriate class and call methods there to handle renewals, events, etc.

I'm thinking that the Membership and MembershipTerm entities should not really have that much logic...

I've also created 2 new Exception classes (so far): MembershipFeatureNotImplementedException (for membership types that don't support a particular method) and MembershipNotFoundException (for membership retrieval methods that don't find a membership for a given user).

**Comment by freelock** [comment]
(race conditions!)

> I think we need to flesh out the exact purpose of the "Membership Term," as you've worked it up so far.
>
Ok.
 
> Simply put, I think there's some conflict in approach between the state of the membership, and the state of the membership term. When I originally wrote the module, the "Membership Offer," which you're replacing, was a one-to-many relationship to memberships which implemented that offer. Basically the membership itself had a state, and the offer represented a particular set of terms or benefits.
> 
> You've taken the "Membership Term" concept from the D7 module and brought it here, where (I believe) you expect a 1-to-1 relationship to Memberships.

No, not 1-to-1, 1 to many...

>  From my read of the D7 membership_entity module, the piece that is so far missing is that a Membership is 1-to-many to membership terms, which might represent re-joins, etc. Your PR has a workflow state on the Membership Term but does not remove the state of the Membership - is that purposeful or an oversight? 

That is purposeful...

> I still think it's the Membership itself that has a state. That state would then change in response to the Membership Term, e.g., one month, and a Membership with a month-long term could be changed to expiration via some sort of audit method (which doesn't appear to exist currently.)
> 

Agreed. (I haven't yet discovered the best way to trigger an event like this based on a time -- do you have something handy to point me to?)

In my thinking, a "Membership" is something that a user has exactly one of (of a given type -- totally fine to have another membership of a different type). The Membership entity contains data about the membership itself -- the member number, the access granted, additional user accounts associated with a membership. It is generally either "Active" (which grants access) or some variation of "Inactive" - pending, expired -- which blocks the associated access.

A "MembershipTerm" defines the period of access, and has its own state -- pending, active, expired. A membership might have several membership terms associated, some in the past/expired, 1 currently active, 1 or more future terms that are not yet active, and may or may not be paid for. The same membership might have terms with different lengths associated. 

In general I would think we would define some default constraints here:

- a user may only have one membership of a specified type
- the membership type may define a user reference field to make a membership "belong" to multiple users -- not necessarily just the author
- a membership may only have one active term (but may have any number of terms in other states)
- a MembershipTermType is associated with only one MembershipType (but there may be many MembershipTermTypes for a given MembershipType)
- a MembershipTerm is associated with only one Membership

... and I can certainly see allowing these constraints to be overridden by other MembershipType/MembershipTermTypes if a site needs that.


> If the Membership Term entity is merely a way of establishing and tracking the time-based term of membership, your note in the issue queue about using inline entity form probably becomes the better approach, and we can remove any of the term's own HTML routes. The Membership Offer entity was built with display in mind because it represented something independent of a specific membership, but a Membership Term doesn't appear to do that.

Hmm. Good thinking, we probably don't need direct access.

I am thinking that there would be membership products that would link to a particular MembershipTermType -- so when that product is purchased, it creates a new MembershipTerm entity and adds it to a user's existing membership, or creates a new membership of the type defined by the MembershipTermType.

So for example, you might have a "Basic" and "Premium" membership types, and "Basic- 1year", "Basic- 3month", "Premium - 1year", "Premium - 3month", "Premium - 1month" membership term types. When you purchase a product that is for a "Premium - 1month" membership term, for a specific user, there's a bunch of scenarios the membership module might facilitate:

### New membership

- Create new membership of type "Premium"
- Create new membership term of type "Premium - 1month" linked to the new membership, with start date of today and end date of today +1 month

### Renewal of "Expiring" membership

- Retrieve the "Premium" membership associated with the user
- Retrieve the end date of the last membership term
- Create a new membership term of type "Premium - 1month" linked to the membership, with a start date matching the previous end date, and an end date of that date +1 month

### Renewal of "Expired" membership

In this scenario, the difference from the previous scenario might be that the new membership starts today, not the end date of the previous membership term.

## Other considerations:

- Our current client wants to sell essentially the same membership term types at different prices, self-selected by the users. (e.g. they charge professionals with a certification more than they charge students for the same benefits on the site) -- suggests a need for a separate entity for the product and the membership term (0 to many products per membership term)

- different workflows available for different MembershipTermTypes -- for example, (assuming workflow is a reasonable place to model this), an annual membership may have several more states than a monthly, with state changes triggering different email notifications. I'm thinking state is useful here, because an early payment could easily skip several of those notification states.

So I'm picturing that when you create a new MembershipTermType, you select a MembershipType, a Workflow, a term length (including lifetime) and a class implementing MembershipTypeInterface (for allowing alternative business processes that need code to properly implement), and the result is something that Commerce (or other systems) can just ask the membership.manager service for the registration form for a new MembershipTerm of this type, and it comes back with optional fields necessary for the MembershipTerm and/or the corresponding Membership... a different method could then create the new Membership and term in one go.

Cheers,
John

**Comment by freelock** [comment]
P.S. TLDR on the membership state vs membership term state:

Membership state defines whether or not a user's membership is active, and emits events when this changes to grant/deny access to whatever relies upon this membership.

MembershipTerm state is more about workflow around the membership renewal process, and might leverage different state transitions to trigger notifications.

These states will likely also interact with each other.

**Comment by bradjones1** [comment]
OK... A lot here, so let me distill this down to some near-term conclusions and to-dos.

## The Membership entity

It sounds like we have concluded a membership is the core functionality, a record of the member's status. In your use case, users would have only one membership, but from an architecture standpoint I think we can be agnostic on this point and enforce that via add-on functionality, e.g. an entity validation check in contrib or custom code.

Memberships have a bundle which is primarily used to define the State Machine workflow for the membership. Additional fields would pertain to site-specific business logic.

### Membership state and transitions

The membership entity has a state, which operates within a workflow defined on the Membership Type (bundle). State machine emits events relating to state transitions.

## Managing Memberships: "Terms," "Providers," etc.

The more I think about this, the more it sounds like the core module should define Membership and MembershipType (bundle) content and config entities only, and leave the rest up to either bundled, optional submodules or contrib. You and I are coming at this problem from different use cases but it comes down to this: **Memberships are assumed to be created and managed via an external entity or module that triggers the Membership's state transitions.**

This is what you've re-worked from the original "Offer" concept to a "Term," in line with D7 Membership Entity, and what I've done with "Membership Providers." From a very high level, however, both do the same thing. The Membership Term starts (creating a Membership), nears expiration (prompting an "expiring" state) and ends (ending the Membership via another state transition.) The "provider" model is nearly identical minus the requirement for a Drupal entity to record state; this is just done on an external system. In effect, the Membership Term is remotely stored.

From an architectural standpoint, the only real requirement for the Membership to provide a reference to the "thing" that is responsible for its management, e.g., the Term. In the case of an external provider, perhaps that "thing" is a plugin instance.

That sounds like a great job for an Interface, which could be generic enough to be implemented by both entities and plugins, and has methods like `getProvider()` (the module name) or `supportsExpiring()`, etc. Or perhaps there's a base Interface and then more specific Interfaces for different general types of membership management.

## Near-term to-dos

If that sounds good, I propose we make a near-term goal of:

1. separating out the Membership Term functionality into a separate module (that maybe you could be lead maintainer on) and
1. codifying the Membership - Membership Type - Provider model with fields and methods on Membership, as well as defining the minimum qualities of a module that provides for membership management, and making both the existing Membership Provider and your Membership Term modules reference implementations for a plugin-driven and entity-driven model, respectively.

**Comment by freelock** [comment]
Hi, Brad,

If you'd like to chat more real time, you can reach me at https://matrix.to/#/@john:matrix.freelock.com -- Riot provides a persistant chat, even if I'm not around I'll get a message...

Comments:

>  In your use case, users would have only one membership, but from an architecture standpoint I think we can be agnostic on this point and enforce that via add-on functionality, e.g. an entity validation check in contrib or custom code.

Hmm. I'm curious about what scenarios/requirements would necessitate allowing a user to have more than one membership of a particular type. Do you have a good example?

I was thinking of taking the opposite approach -- build the constraint/validation check into the base module, but allow a sub-class to override this constraint and allow multiple memberships per type per user.

In either case, we need to define the interface(s) to support either case...

> From an architectural standpoint, the only real requirement for the Membership to provide a reference to the "thing" that is responsible for its management, e.g., the Term. In the case of an external provider, perhaps that "thing" is a plugin instance.

Hmm yes, sounds pretty workable, and is the direction I've been coding (although so far my work has been explicitly MembershipTerms, so I should generalize that).

> That sounds like a great job for an Interface, which could be generic enough to be implemented by both entities and plugins, and has methods like getProvider() (the module name) or supportsExpiring(), etc. Or perhaps there's a base Interface and then more specific Interfaces for different general types of membership management.

Mostly agreed -- I'm going down the approach of creating a MembershipManager service that will do much of this, but I think having a well-defined interface for the Membership and whatever the managing entity is would be great.

I am thinking that we can (should) define an interface that has a broad range of methods available for a large number of common tasks, and in a base class, throw a MembershipFeatureNotImplemented exception for any that are not defined for a class.

I'm mainly thinking of ecosystem development here -- I'd like to see a coherent API, and allow for a broad range of pluggable classes that provide different features... callers can catch the exception and turn off unsupported features while still having a single method to call that other membership types might implement.

So yes, I think the big role of a Membership API module is... to provide an API. If you don't think we should ship an actual working membership module...

> If that sounds good, I propose we make a near-term goal of:

> 1. separating out the Membership Term functionality into a separate module (that maybe you could be lead maintainer on) and

Hmm. I'd be happy to maintain a separate module for this -- but I do think it belongs in the Membership module, at least as a submodule. Mainly so ecosystem developers don't go off and implement lots of duplicate functionality and fragment the API.

> 2. codifying the Membership - Membership Type - Provider model with fields and methods on Membership, as well as defining the minimum qualities of a module that provides for membership management, and making both the existing Membership Provider and your Membership Term modules reference implementations for a plugin-driven and entity-driven model, respectively.

Agreed. I'm nearing having functioning code for this now, can refactor to make more generalized.

**Pull request closed without merging**